Firestore/core/test/firebase/firestore/remote/grpc_stream_test.cc

```diff
@@ -102,157 +147,409 @@ class GrpcStreamTest : public testing::Test {
+// an error in gRPC:
+// CannotWriteAfterWriteLast
+// CannotWriteLastAfterWriteLast
+
+// Read and write
+
 TEST_F(GrpcStreamTest, ReadIsAutomaticallyReadded) {
-  StartStream();
-  ForceFinish({/*Read*/ Ok});
-  EXPECT_EQ(observed_states(), States({"OnStreamStart", "OnStreamRead"}));
+  worker_queue.EnqueueBlocking([&] { stream->Start(); });
 
-  ForceFinish({/*Read*/ Ok});
-  EXPECT_EQ(observed_states(),
-            States({"OnStreamStart", "OnStreamRead", "OnStreamRead"}));
+  ForceFinish({{Type::Read, MakeByteBuffer("foo")}});
+  EXPECT_EQ(observed_states(), States({"OnStreamStart", "OnStreamRead(foo)"}));
+
+  ForceFinish({{Type::Read, MakeByteBuffer("bar")}});
+  EXPECT_EQ(observed_states(), States({"OnStreamStart", "OnStreamRead(foo)",
+                                       "OnStreamRead(bar)"}));
 }
 
 TEST_F(GrpcStreamTest, CanAddSeveralWrites) {
-  StartStream();
+  worker_queue.EnqueueBlocking([&] { stream->Start(); });
 
   worker_queue.EnqueueBlocking([&] {
-    stream().Write({});
-    stream().Write({});
-    stream().Write({});
+    stream->Write({});
+    stream->Write({});
+    stream->Write({});
+  });
+
+  int reads = 0;
+  int writes = 0;
+  ForceFinish([&](GrpcCompletion* completion) {
+    switch (completion->type()) {
+      case Type::Read:
+        ++reads;
+        completion->Complete(true);
+        break;
+      case Type::Write:
+        ++writes;
+        completion->Complete(true);
+        break;
+      default:
+        EXPECT_TRUE(false) << "Unexpected completion type "
+                           << static_cast<int>(completion->type());
+        break;
+    }
+
+    bool done = writes == 3;
+    return done;
   });
-  ForceFinish({/*Read*/ Ok, /*Write*/ Ok, /*Read*/ Ok, /*Write*/ Ok,
-               /*Read*/ Ok, /*Write*/ Ok});
 
-  EXPECT_EQ(observed_states(), States({"OnStreamStart", "OnStreamRead",
-                                       "OnStreamRead", "OnStreamRead"}));
+  EXPECT_EQ(writes, 3);
+  EXPECT_EQ(observed_states().size(), reads + /*Start*/ 1);
+  EXPECT_EQ(observed_states().back(), "OnStreamRead");
 }
 
+// Observer
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnStart) {
+  worker_queue.EnqueueBlocking([&] { stream->Start(); });
+  // `Start` is a synchronous operation.
+  EXPECT_EQ(observed_states(), States({"OnStreamStart"}));
+}
+
+// `ObserverReceivesOnRead` is tested in `ReadIsAutomaticallyReadded`
+
 TEST_F(GrpcStreamTest, ObserverReceivesOnError) {
-  StartStream();
+  worker_queue.EnqueueBlocking([&] { stream->Start(); });
+
+  ForceFinish({{Type::Read, Error}});
+  // Give `GrpcStream` a chance to enqueue a finish operation
+  ForceFinish({{Type::Finish, grpc::Status{grpc::RESOURCE_EXHAUSTED, ""}}});
 
-  // Fail the read, but allow the rest to succeed.
-  ForceFinish({/*Read*/ Error});  // Will put a "Finish" operation on the queue
+  EXPECT_EQ(observed_states(),
+            States({"OnStreamStart", "OnStreamFinish(ResourceExhausted)"}));
```

Comment:
I might misunderstand the question, but IIUC, it's about why line 374 (`ForceFinish({{Type::Read, Error}});`) doesn't produce a notification. When an operation fails, the stream doesn't notify the observer -- instead, it initiates a "finish" operation and only notifies the observer once the "finish" operation succeeds (so that it can also report the error) -- [see here](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/remote/grpc_stream.cc#L313-L319). So this is by design -- `OnRead` is only ever invoked with successful reads.